packages/core/src/api/blockManipulation/commands/moveBlocks/moveBlocks.ts

```diff
@@ -0,0 +1,309 @@
+import { NodeSelection, Selection, TextSelection } from "prosemirror-state";
+import { CellSelection } from "prosemirror-tables";
+
+import { Block } from "../../../../blocks/defaultBlocks.js";
+import type { BlockNoteEditor } from "../../../../editor/BlockNoteEditor";
+import { BlockIdentifier } from "../../../../schema/index.js";
+import { getNearestBlockPos } from "../../../getBlockInfoFromPos.js";
+import { getNodeById } from "../../../nodeUtil.js";
+
+type BlockSelectionData = (
+  | {
+      type: "text";
+      headBlockId: string;
+      anchorOffset: number;
+      headOffset: number;
+    }
+  | {
+      type: "node";
+    }
+  | {
+      type: "cell";
+      anchorCellOffset: number;
+      headCellOffset: number;
+    }
+) & {
+  anchorBlockId: string;
+};
+
+// `getBlockSelectionData` and `updateBlockSelectionFromData` are used to save
+// and restore the selection within a block, when the block is moved. This is
+// done by first saving the offsets of the anchor and head from the before
+// positions of their surrounding blocks, as well as the IDs of those blocks. We
+// can then recreate the selection by finding the blocks with those IDs, getting
+// their before positions, and adding the offsets to those positions.
+function getBlockSelectionData(
+  editor: BlockNoteEditor<any, any, any>
+): BlockSelectionData {
+  const state = editor._tiptapEditor.state;
+  const selection = state.selection;
+
+  const anchorBlockPosInfo = getNearestBlockPos(state.doc, selection.anchor);
+
+  if (selection instanceof CellSelection) {
+    return {
+      type: "cell" as const,
+      anchorBlockId: anchorBlockPosInfo.node.attrs.id,
+      anchorCellOffset:
+        selection.$anchorCell.pos - anchorBlockPosInfo.posBeforeNode,
+      headCellOffset:
+        selection.$headCell.pos - anchorBlockPosInfo.posBeforeNode,
+    };
+  } else if (editor._tiptapEditor.state.selection instanceof NodeSelection) {
+    return {
+      type: "node" as const,
+      anchorBlockId: anchorBlockPosInfo.node.attrs.id,
+    };
+  } else {
+    const headBlockPosInfo = getNearestBlockPos(state.doc, selection.head);
+
+    return {
+      type: "text" as const,
+      anchorBlockId: anchorBlockPosInfo.node.attrs.id,
+      headBlockId: headBlockPosInfo.node.attrs.id,
+      anchorOffset: selection.anchor - anchorBlockPosInfo.posBeforeNode,
+      headOffset: selection.head - headBlockPosInfo.posBeforeNode,
+    };
+  }
+}
+
+// `getBlockSelectionData` and `updateBlockSelectionFromData` are used to save
```

Comment:
let's use jsdoc for longer method comments?

Maybe we can use https://github.com/gajus/eslint-plugin-jsdoc/blob/main/docs/rules/convert-to-jsdoc-comments.md to enforce this as a rule, and to convert the codebase? (can also be done separately)